File: 프로그래머스/2/42587. 프로세스/프로세스.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
#include <iostream>
using namespace std;

int solution(vector<int> priorities, int location) {
    queue<pair<int,int>> q; // {우선순위, 원래 인덱스}
    priority_queue<int> mx; // 현재 큐의 최대 우선순위
    int targetNumber = priorities[location]; // 찾으려는 숫자의 위치
    int maxNumber; // 가장 큰 수
    
    for(int i=0; i<priorities.size(); i++) { 
        q.push({priorities[i], i});
        mx.push(priorities[i]);
    }
    maxNumber = mx.top();
     
    int i = 0;
    while(true) {
        int pr = q.front().first;
        int idx = q.front().second;
        
        if(pr == maxNumber) {
            q.pop();
            mx.pop();
            i++;
            if(idx == location) {
                return i;
            }
            if (!mx.empty()) {
                maxNumber = mx.top();
            }
        } else {
            q.push(q.front());
            q.pop();
        }
    }
}

File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.*;
class Solution {
    public int solution(int[] priorities, int location) {        
        Queue<int[]> q = new LinkedList<>();
        for(int i=0; i<priorities.length; i++) {
            q.add(new int[]{i, priorities[i]});
        }
        
        int rank = 1;
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            boolean flag = false;
            
            for(int[] x : q) {
                if(x[1] > cur[1]) {
                    flag = true;
                    break;
                }
            }
            if(flag) {
                q.add(new int[]{cur[0], cur[1]});
            } else {
                if(cur[0] == location) {
                    return rank;
                }
                rank++;
            }
        }
        return rank;
    }
}